Various fixes to backup solution

NoxHarmonium opened the pull request:
We were having some issues getting [the AWS reference efs-to-efs backup solution](https://docs.aws.amazon.com/solutions/latest/efs-to-efs-backup/welcome.html) to work out of the box in our AWS environment so we made some fixes to get it working that can be shared with the wider community.

I hope these fixes can help others!

**New Features**
- Adds support for HTTP proxies
- Adds support for adding a keypair to the backup EC2 instance
- Can now use three mount points for the backup EFS volume
- Can attach a custom security group to the backup EC2 instance
- Fix issue where backup script would exit before reporting finished state to orchestrator
- Made backup script more robust
- Misc. cleanup and various small fixes

**Still Needs To Be Done**
- I couldn't get the tests to work but I don't have time to debug that at the moment
- I haven't looked at the restore scripts yet, they probably won't work with a proxy. I'll have to look at that later.

NoxHarmonium committed:
Prevent child from exiting the parent script early

The parent script needs to shut down the instance. If exit is called at this stage the instance is never terminated.

NoxHarmonium committed:
Manually fetch AWS secrets from instance profile

NoxHarmonium committed:
Ensure that the EC2 instance is stopped on error

The script should not keep running if there is an error. It should stop immediately and then shut down the EC2 instance so the error can be handled before the time window has elapsed.

NoxHarmonium committed:
Fix invalid URL

NoxHarmonium committed:
Fix the URL properly

I only changed the URL in the echo statement, not the actual command.

NoxHarmonium committed:
Ensure that CURL causes error if status not 200

NoxHarmonium committed:
Minor code cleanup

NoxHarmonium committed:
Remove commented code

NoxHarmonium committed:
Add time to log

NoxHarmonium committed:
Ignore error from `rm` if file doesn't exist

Commit:
Allow attaching a custom SG to an instance

- With this change you can create an SG that allows access to the source
EFS target and then specify it in the template so it is automatically
attached to the backup instance when building the solution.

Commit:
Allow 3 subnets to be chosen for destination EFS

- There doesn't seem to be any disadvantages of having mount targets in
3 availability zones even if you don't have instances in each one. I
think 3 availability zones seems like a more common arrangement.
- If the number of mount targets is something that varies a lot between
different systems we might have to create 3 separate templates because
it doesn't look like we can generate mount point resources dynamically
depending on how many subnets are chosen.

Commit:
Enable proxy support for backup solution

- In the environment we are installing the backup solution in, we do not
have internet access except through a proxy.
- This adds optional support for specifying a proxy in the template.
- The SSM agent needed the proxy configuration to be added to its init
script which is a bit ugly but works fine for now.

Commit:
Allow specifying custom script URLS

- This allows a developer to point the solution at different scripts for
debugging purposes. It should default to the standard ones.

Commit:
Add keypair parameter to template

- This allows a developer to log onto the backup instance with SSH to
diagnose any issues with the backup instance.

Commit:
Specify parameter type

- This makes it much easier to choose an SG when setting up the template

pearj commented on source/scripts/efs-backup-fpsync.sh:
```diff
@@ -36,8 +36,8 @@ sudo yum -y install nfs-utils
-echo "-- $(date -u +%FT%T) -- wget https://s3.amazonaws.com/%TEMPLATE_BUCKET_NAME%/efs-backup/latest/fpart.zip"
-wget https://s3.amazonaws.com/%TEMPLATE_BUCKET_NAME%/efs-backup/latest/fpart.zip
+echo "-- $(date -u +%FT%T) -- wget https://s3.amazonaws.com/solutions-reference/efs-backup/latest/fpart.zip"
+wget https://s3.amazonaws.com/solutions-reference/efs-backup/latest/fpart.zip
```
Probably want to keep `%TEMPLATE_BUCKET_NAME%` here

NoxHarmonium commented:
I just realised that while I was working on this, a change went through to master which will probably conflict with this PR. I will have to rebase and merge my changes in.

gsingh04 commented:
thanks for providing the PR, we will look to incorporate these in v1.2

NoxHarmonium commented:
Thanks for the update @gsingh04! 

I have just rebased my changes on top of master and removed some redundant commits. I'm wondering if I should remove commit 2a0f9ad460a2065d98489013ffb3a30a0a734501 because it might have just been a problem with our environment with all the proxy requirements etc. I will need someone else to verify in their environment.

gsingh04 commented:
@NoxHarmonium verified 2a0f9ad, worked fine for me. Also, proxies should not be a problem when pulling instance metadata information. Did you make sure your instance had a IAM role attached to it when running the script?

Commit:
Fix template variable that was replaced

Commit:
Remove code that should not be needed on a normal EC2 instance

- The AWS management API IP (169.254.169.254) needs to be added to
no_proxy for the AWS tool to work properly. Once the management API IP
address was added to no_proxy this code was no longer needed.

NoxHarmonium commented:
@gsingh04 I triple checked that I had the correct IAM role attached to the instance but I still wasn't able to get it to work. It turned out to be that I did not put the AWS management API IP address into the "no_proxy" environment variable.

This requirement is outlined in the AWS documentation at https://docs.aws.amazon.com/cli/latest/userguide/cli-http-proxy.html#cli-http-proxy-ec2

I have removed that block of code because it should not be needed. I have also fixed up the removed template variables mentioned by @pearj.

quiver commented:
- Adds support for adding a keypair to the backup EC2 instance
- Can now use three mount points for the backup EFS volume

+1

EFS mount points are really important.

AWS recommends to create mount points for every AZ but CFn template creates only two.

Imagine this scenario.

If you specify more than two subsets as CFn parameters, 
- Auto Scaling Group uses all of them
- EFS mount points use two of them

If instances are launched from non-supported AZ, backup job fails.

It took me a while to debug why EFS-to-EFS solution fails randomly  :-(

gsingh04 commented:
Thanks for your feedback. Please try our latest v1.3 release and let us know the performance improvement or if there are any concerns. 
- Adding key pair is a change that can be done at customer end. We do not recommend adding key pair as we provide Cloudwatch dashboard and logs in S3 to debug issues.
- Some of the regions have only 2 AZs currently, like Mumbai, Canada. EFS is not available in these regions for now, but we did not want the solution to break once EFS is available in these regions; hence the decision to go with 2 AZs
